Make the ipfs transport extensible.

cobward opened the pull request:
I've implemented two methods to do this.

The first is with the `build_transport` and `build_transport_with_upgrade` methods. The latter of those creates a ipfs-compatible transport with custom transport layers and a single custom upgrade.

The second method is within the `apply_upgrades` submodule. This contains a macro (apply_upgrades) which allows the user to apply any number of upgrades when building the transport.

The first method is I think clearer and easier to use, however it doesn't allow more than one upgrade to be applies. The pros and cons are switched for the second method. Thoughts?

cobward committed:
Make the ipfs transport extensible.

chunningham commented:
IMO yeah agree the first method is clearer, overall I like the approach. Extending it to apply multiple upgrades I guess has typing issues that make the macro appropriate. Could the `TransportUpgrader` wrap an `upgrade::Authenticated` to allow `.apply` to be called several times in a builder style without a macro? then `TransportBuilder.then` calls `.authenticate`. maybe this faces similar type issues in practice

cobward commented:
I did try that and I think it introduced more issues with trait bounds. Will give it another go now because I can't remember why I gave up on it.

cobward commented:
Ah yeah I gave up on it yesterday because I couldn't figure out how to write the type `T` inside `Authenticated<T>`. With fresh eyes, and a bit of trial and error, I've figured it out. Thanks for the prompt on this!

cobward committed:
More ergonomical method for multiple upgrades.

cobward commented:
I've realised that for kepler we actually need to add upgrades pre-authentication, and there isn't a way to chain upgrades in this way in libp2p. So tomorrow I'll work on a solution to this, which I have a decent idea of.

chunningham commented:
It seems like the "handshake" might require an implementation of in/outbound upgrades which satisfy the `U` bound on `Builder::authenticate`, which essentially augments `NoiseAuthenticated` (result of `NoiseConfig::into_authenticated`) with extra steps? to recap, my understanding is that we want to:

* establish a noise-encrypted channel (`NoiseAuthenticated` impls of in/outbound upgrade)
* use said channel to exchange authz proof, close channel if proof insufficient

Is it the case that the "per-connection" logic that gets actioned in an "handshake" way is the upgrade passed in to `Builder::authenticate`, so acts differently from the upgrades applied to `Authenticated` which are selected from when negotiating a protocol upgrade?

cobward committed:
Make it possible to map the output of auth.

This would allow a user to implement additional checks on the peer id
(such as checking this peer is authorized to connect) before
establishing a connection.

cobward commented:
Yep that's pretty much what I've done, except:

> use said channel to exchange authz proof

I haven't implemented it to be able to use the channel. For our purposes I think we don't need to communicate over the channel, we just need to establish whether the inbound PeerId is authorized to be part of the orbit. I may be misunderstanding how libp2p treats an error raised on the listening side however

chunningham commented on src/p2p/mod.rs:
```diff
@@ -61,7 +61,7 @@ pub async fn create_swarm<TIpfsTypes: IpfsTypes>(
 
     // Set up an encrypted TCP transport over the Mplex protocol.
-    let transport = transport::build_transport(options.keypair.clone())?;
+    let transport = transport::TransportBuilder::new(options.keypair.clone())?.build();
```
can we add a transport arg (either builder or `TTransport`) into this fn and `UninitializedIpfs::start` so it can be swapped out?

cobward replied:
I've got a version of kepler-staging in my fork where I've consolidated these changes with the behaviour and other changes for kepler. In particular this commit does as you suggest:
https://github.com/cobward/rust-ipfs/commit/9f2b170c0cf908442761763026a01629cfb06bee

In terms of putting it on this PR, I think we will want to have a commit that breaks open the UninitialisedIpfs, which both the behaviour change and transport change can be based on, or else raise both changes in a single PR.

cobward replied:
It might make sense to do both in a single PR since there is some interdependency (e.g. relay transport requires adding relay to the behaviour), and they are in a similar spirit of making the underlying p2p behaviour extensible.

chunningham replied:
if they can be ordered or separated into PRs it would be easier for the reviewer but I see your point w.r.t. the interdependency

cobward committed:
tmp test
